web/src/util/serialization/apd.js

```diff
@@ -0,0 +1,192 @@
+ */
+export const toAPI = (
+  apdState,
+  activityState,
+  serializeActivity = activityToAPI
+) => {
+  const incentivePayments = Object.entries(
+    apdState.incentivePayments.ehAmt
+  ).map(([year, quarters]) => ({
+    year,
+    q1: {
+      ehPayment: quarters[1],
+      ehCount: apdState.incentivePayments.ehCt[year][1],
+      epPayment: apdState.incentivePayments.epAmt[year][1],
+      epCount: apdState.incentivePayments.epCt[year][1]
+    },
+    q2: {
+      ehPayment: quarters[2],
+      ehCount: apdState.incentivePayments.ehCt[year][2],
+      epPayment: apdState.incentivePayments.epAmt[year][2],
+      epCount: apdState.incentivePayments.epCt[year][2]
+    },
+    q3: {
+      ehPayment: quarters[3],
+      ehCount: apdState.incentivePayments.ehCt[year][3],
+      epPayment: apdState.incentivePayments.epAmt[year][3],
+      epCount: apdState.incentivePayments.epCt[year][3]
+    },
+    q4: {
+      ehPayment: quarters[4],
+      ehCount: apdState.incentivePayments.ehCt[year][4],
+      epPayment: apdState.incentivePayments.epAmt[year][4],
+      epCount: apdState.incentivePayments.epCt[year][4]
+    }
+  }));
+
+  const apd = {
+    activities: Object.values(activityState.byKey).map(serializeActivity),
+    federalCitations: apdState.assurancesAndCompliance,
+    incentivePayments,
+    narrativeHIE: apdState.hieNarrative,
+    narrativeHIT: apdState.hitNarrative,
+    narrativeMMIS: apdState.mmisNarrative,
+    pointsOfContact: apdState.pointsOfContact,
+    previousActivityExpenses: Object.entries(
+      apdState.previousActivityExpenses
+    ).map(([year, o]) => ({
+      year,
+      hie: o.hie,
+      hit: o.hit,
+      mmis: o.mmis
+    })),
+    previousActivitySummary: apdState.previousActivitySummary,
+    programOverview: apdState.overview,
+    stateProfile: apdState.stateProfile,
+    years: apdState.years
+  };
+
+  return apd;
+};
+
+/**
+ * Deserialize an APD object from the API into an object
+ * matching redux state shape.
+ * @param {*} apdAPI - APD object from the API
+ * @param {Object} defaults - default values for missing properties
+ * @param {Object} defaults.activities - default activities
+ * @param {Object} defaults.assurancesAndCompliance - default assurances and compliance
+ * @param {Object} defaults.hieNarrative - default HIE narrative
+ * @param {Object} defaults.hitNarrative - default HIT narrative
+ * @param {Object} defaults.incentivePayments - default value for incentive payments
+ * @param {Object} defaults.mmisNarrative - default MMIS narrative
+ * @param {Object} defaults.overview - default APD overview
+ * @param {Object} defaults.pointsOfContact - default points of contact
+ * @param {Object} defaults.previousActivityExpenses - default value for previousActivityExpenses
+ * @param {Object} defaults.previousActivitySummary - default activity summary
+ * @param {Object} defaults.stateProfile - default state profile
+ * @param {Object} defaults.years - default value for years
+ */
```

Comment:
Added this JSDoc because it turns out VSCode is really good at reading it and turning it into a decently useable tooltip.

Reply:
whoa, cool